File: Data Structure/Linked List/Medium/LET_no143_Reorder List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head:
            return
        
        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
            
        last = None
        cur = slow.next
        slow.next = None
        
        while cur:
            nxt = cur.next
            cur.next = last
            last = cur
            cur = nxt
        
        head2 = head
        while last:
            nxt = head2.next
            head2.next = last
            head2 = head2.next
            last = nxt
        

File: Data Structure/Linked List/Medium/LET_no143_Reorder List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        if (!head)
            return;
        
        ListNode *slow = head, *fast = head;
        while (fast->next && fast->next->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        
        ListNode *cur = slow->next, *last = nullptr;
        slow->next = nullptr;
        
        while (cur) {
            ListNode *nxt = cur->next;
            cur->next = last;
            last = cur;
            cur = nxt;
        }
        
        ListNode *head2 = head;
        while (last) {
            ListNode *nxt = head2->next;
            head2->next = last;
            head2 = head2->next;
            last = nxt;
        }
    }
};